rispy/parser.py

```diff
@@ -120,226 +95,175 @@ def __init__(
-                    self.last_tag = None
-                except NextLine:
+                if tag in self.ignore:
                     continue
-            else:
-                try:
-                    yield self._parse_other(line)
-                except NextLine:
+
+                if tag == self.END_TAG:
+                    result.append(record)
+
+                    record = self._iter_till_start(lines)
                     continue
 
-    def _parse_tag(self, line):
-        tag = self.get_tag(line)
-        if tag in self.ignore:
-            raise NextLine
-
-        if tag == self.END_TAG:
-            return self.current
-
-        if tag == self.START_TAG:
-            # New entry
-            if self.in_ref:
-                raise ParseError(f"Missing end of record tag in line {self.line_number}:\n {line}")
-            self._add_tag(tag, line)
-            self.in_ref = True
-            raise NextLine
-
-        if not self.in_ref:
-            raise ParseError(f"Invalid start tag in line {self.line_number}:\n {line}")
-
-        if tag in self.mapping:
-            self._add_tag(tag, line)
-            raise NextLine
-        elif not self.skip_unknown_tags:
-            self._add_unknown_tag(tag, line)
-            raise NextLine
-
-        raise NextLine
-
-    def _parse_other(self, line):
-        if self.skip_missing_tags:
-            raise NextLine
-        if self.in_ref:
-            # Active reference
-            if self.last_tag is None:
-                raise ParseError(f"Expected tag in line {self.line_number}:\n {line}")
-            # Active tag
-            self._add_tag(self.last_tag, line, all_line=True)
-            raise NextLine
-
-        if self.is_header(line):
-            raise NextLine
-        raise ParseError(f"Expected start tag in line {self.line_number}:\n {line}")
-
-    def _add_single_value(self, name, value, is_multi=False):
+                self._add_tag(record, tag, content)
+                last_tag = tag
+
+        except StopIteration:
+            return result
+
+    def parse_line(self, line):
+        """Parse line of RIS file.
+
+        This method parses a line between the start and end tag.
+        It returns the tag and the content of the line. Typically,
+        the first 2 characters are the tag, followed by a seperator,
+        and the rest of the line is the content.
+
+        Custom parsers can override this method to change the way
+        lines are parsed. For example, a very basic RIS parser would
+        return the first 2 characters as the tag and the rest of the
+        line as the content of the tag. `(line[0:2], line[6:].strip())`
+
+        Parameters
+        ----------
+        line : str
+            Line of RIS file between start and end tag.
+
+        Returns
+        -------
+        tuple
+            Tuple containing the tag and the content of the tag.
+        """
+        if line[3:5] == " -" and line[:2].isupper():
```

Comment:
This looks good, though you might as well test `line[2:5] == "  -"` here, right? That should have a lower chance of false positives for multiline content.

Basically this line is where all the parsing magic happen, and you can do it with string slicing or with regex. I ran a small test to see how these two compare:

```[python]
import regex

pattern = regex.compile(r"^[A-Z][A-Z0-9]  -")


def parse_with_regex(s):
    if bool(pattern.match(s)):
        return s[:2]


def parse_with_slice(s):
    if s[2:5] == "  -" and s[:2].isupper():
        return s[:2]


def test_parse_with_regex_match(benchmark):
    s = "AB  - this is a random text"
    assert benchmark(parse_with_regex, s) == "AB"


def test_parse_with_regex_no_match(benchmark):
    s = "this is a random text"
    assert benchmark(parse_with_regex, s) is None


def test_parse_with_slice_match(benchmark):
    s = "AB  - this is a random text"
    assert benchmark(parse_with_slice, s) == "AB"


def test_parse_with_slice_no_match(benchmark):
    s = "this is a random text"
    assert benchmark(parse_with_slice, s) is None
```

This gave the following results:
```
----------------------------------------------------------------------------------------------- benchmark: 4 tests ----------------------------------------------------------------------------------------------
Name (time in ns)                       Min                    Max                Mean              StdDev              Median                IQR            Outliers  OPS (Mops/s)            Rounds  Iterations
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_parse_with_slice_no_match     124.0000 (1.0)       1,167.0000 (1.0)      138.3356 (1.0)       31.6514 (1.0)      132.0000 (1.0)       7.0000 (1.33)    1885;2031        7.2288 (1.0)       76924         100
test_parse_with_slice_match        257.8947 (2.08)      7,605.2631 (6.52)     291.6303 (2.11)     101.2712 (3.20)     284.2105 (2.15)      5.2632 (1.0)    2315;22931        3.4290 (0.47)     196079          19
test_parse_with_regex_no_match     314.9999 (2.54)      6,020.0000 (5.16)     356.7073 (2.58)     119.1388 (3.76)     345.0000 (2.61)     15.0000 (2.85)    2646;3596        2.8034 (0.39)     149254          20
test_parse_with_regex_match        399.9994 (3.23)     28,300.0008 (24.25)    542.0298 (3.92)     359.8332 (11.37)    500.0002 (3.79)     99.9989 (19.00)       66;80        1.8449 (0.26)      34723           1
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------


Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
```

So string slicing seems the clear winner!

Reply:
Actually, the logic still isn't correct. This implementation will accept `1A  -` as a valid tag even though a digit is not allowed as the first character. You'd need something like `line[2:5] == "  -" and line[:2].isupper() and not line[:1].isdigit()`

Reply:
Resolved